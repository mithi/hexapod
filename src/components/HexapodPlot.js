import React from "react"
import createPlotlyComponent from "react-plotly.js/factory"
import * as defaults from "../templates"
import { getNewPlotParams } from "../hexapod"

class HexapodPlot extends React.Component {
    cameraView = defaults.CAMERA_VIEW
    state = { ready: false }
    Plot = null

    logCameraView = relayoutData => {
        this.cameraView = relayoutData["scene.camera"]
    }

    componentDidMount() {
        import("plotly.js-gl3d-dist-min").then(Plotly => {
            this.Plot = createPlotlyComponent(Plotly)
            this.setState({ ready: true })
        })
    }

    render() {
        if (!this.state.ready) {
            return <p>Loading your cute robot...</p>
        }

        if (!this.props.hexapod) {
            return null
        }
        const [data, layout] = getNewPlotParams(this.props.hexapod, this.cameraView)

        const props = {
            data,
            layout,
            onRelayout: this.logCameraView,
            revision: this.props.revision,
            config: { displaylogo: false, responsive: true },
            style: { height: "100%", width: "100%" },
            useResizeHandler: true,
        }

        const Plot = this.Plot
        return <Plot {...props} />
    }
}

export default HexapodPlot
