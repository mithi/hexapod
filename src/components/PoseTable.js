import React from "react"
import ReactMarkdown from "react-markdown"
import { POSITION_NAMES_LIST } from "../hexapod"

const POSITION_ALIAS = {
    rightMiddle: "rm",
    rightFront: "rf",
    leftFront: "lf",
    leftMiddle: "lm",
    leftBack: "lb",
    rightBack: "rb",
}

const formatPose = pose =>
    POSITION_NAMES_LIST.reduce((formattedPose, position) => {
        const alias = POSITION_ALIAS[position]
        const { alpha, beta, gamma } = pose[position]
        formattedPose[alias] = {
            alpha: Number(alpha).toFixed(2),
            beta: Number(beta).toFixed(2),
            gamma: Number(gamma).toFixed(2),
        }
        return formattedPose
    }, {})

const poseMessage = ({ rm, rf, lf, lm, lb, rb }) => `

| POSITION    | ALPHA       | BETA        | GAMMA       |
|-------------|:-----------:|:-----------:|:-----------:|
| rightMiddle | ${rm.alpha} | ${rm.beta}  | ${rm.gamma} |
| rightFront  | ${rf.alpha} | ${rf.beta}  | ${rf.gamma} |
| leftFront   | ${lf.alpha} | ${lf.beta}  | ${lf.gamma} |
| leftMiddle  | ${lm.alpha} | ${lm.beta}  | ${lm.gamma} |
| leftBack    | ${lb.alpha} | ${lb.beta}  | ${lb.gamma} |
| rightBack   | ${rb.alpha} | ${rb.beta}  | ${rb.gamma} |

`

const poseTable = ({ pose }) => {
    const formattedPose = formatPose(pose)
    const { rm, rf, lf, lm, lb, rb } = formattedPose
    const markdownMessage = poseMessage({ rm, rf, lf, lm, lb, rb })
    return (
        <div className="table-container">
            <div className="cell" style={{}}>
                <ReactMarkdown source={markdownMessage} />
            </div>
        </div>
    )
}

export default poseTable
