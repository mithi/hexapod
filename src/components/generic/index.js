import NumberInputField from "./NumberInputField"
import { BasicButton, ResetButton, Card, ToggleSwitch, AlertBox } from "./SmallWidgets"
import { Slider, sliderList } from "./Slider"
export {
    BasicButton,
    ResetButton,
    Card,
    ToggleSwitch,
    AlertBox,
    NumberInputField,
    Slider,
    sliderList,
}
