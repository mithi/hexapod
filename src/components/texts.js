const SECTION_NAMES = {
    dimensions: "Dimensions",
    inverseKinematics: "Inverse Kinematics",
    forwardKinematics: "Forward Kinematics",
    legPatterns: "Leg Patterns",
    landingPage: "Root",
}

const PATHS = {
    inverseKinematics: {
        path: "/inverse-kinematics",
        description: SECTION_NAMES.inverseKinematics,
    },
    forwardKinematics: {
        path: "/forward-kinematics",
        description: SECTION_NAMES.forwardKinematics,
    },
    legPatterns: { path: "/leg-patterns", description: SECTION_NAMES.legPatterns },
    landingPage: { path: "/", description: SECTION_NAMES.landingPage },
}

const PATH_LINKS = [
    PATHS.inverseKinematics,
    PATHS.forwardKinematics,
    PATHS.legPatterns,
    PATHS.landingPage,
]

const ANGLE_NAMES = ["alpha", "beta", "gamma"]
const DIMENSION_NAMES = ["front", "side", "middle", "coxia", "femur", "tibia"]
const LEG_NAMES = [
    "leftFront",
    "rightFront",
    "leftMiddle",
    "rightMiddle",
    "leftBack",
    "rightBack",
]
const ROTATE_SLIDERS_LABELS = ["rx", "ry", "rz", "hipStance", "legStance"]
const RESET_LABEL = "reset"
const TRANSLATE_SLIDERS_LABELS = ["tx", "ty", "tz"]

/*************
 * NAVIGATION
 *************/

const URLS = {
    KOFI: "https://ko-fi.com/minimithi",
    REPO: "https://github.com/mithi/hexapod",
}

const LINK_DESCRIPTIONS = {
    KOFI: "Buy Mithi Ko-Fi üçµ",
    REPO: "Source Code",
}

/*************
 * LANDING PAGE
 *************/

const LANDING_PAGE_MESSAGE = `

# Mithi's Bare Minimum Hexapod Robot Simulator

- Solve (and visualize) [forward][1] and [inverse][2] kinematics purely on your browser!
It's a complete rewrite of the [one][3] I wrote in Python üêç.
No more server-side computations!

- Consider buying me a [couple cups of coffee üçµ üçµ üçµ][4] to motivate me
to build other robotics related visualizers. (Quadrotors?!)

- Any contribution to improve the source code will be extremely appreciated. üíô
Anything from fixing typographical errors to completely changing the page design ... or even rewriting
modules to follow better software practices. [How you can help][5].

## Love, Mithi

[1]: /forward-kinematics
[2]: /inverse-kinematics
[3]: https://github.com/mithi/hexapod-robot-simulator
[4]: https://ko-fi.com/minimithi
[5]: https://github.com/mithi/hexapod/wiki/Types-of-(code)-Contributions

`

export {
    LANDING_PAGE_MESSAGE,
    SECTION_NAMES,
    ANGLE_NAMES,
    DIMENSION_NAMES,
    LEG_NAMES,
    TRANSLATE_SLIDERS_LABELS,
    ROTATE_SLIDERS_LABELS,
    RESET_LABEL,
    PATHS,
    PATH_LINKS,
    URLS,
    LINK_DESCRIPTIONS,
}
