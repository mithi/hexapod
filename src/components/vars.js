import React from "react"
import { GiCoffeeMug } from "react-icons/gi"
import { FaGithubAlt, FaTimes, FaCheck } from "react-icons/fa"
import { GrStatusGoodSmall } from "react-icons/gr"

const SECTION_NAMES = {
    dimensions: "Dimensions",
    inverseKinematics: "Inverse Kinematics",
    forwardKinematics: "Forward Kinematics",
    legPatterns: "Leg Patterns",
    landingPage: "Root",
}

const PATH_NAMES = {
    inverseKinematics: "/inverse-kinematics",
    forwardKinematics: "/forward-kinematics",
    legPatterns: "/leg-patterns",
    landingPage: "/",
}

const ANGLE_NAMES = ["alpha", "beta", "gamma"]
const DIMENSION_NAMES = ["front", "side", "middle", "coxia", "femur", "tibia"]
const LEG_NAMES = [
    "leftFront",
    "rightFront",
    "leftMiddle",
    "rightMiddle",
    "leftBack",
    "rightBack",
]

const IK_SLIDERS_LABELS = ["tx", "ty", "tz", "rx", "ry", "rz", "hipStance", "legStance"]
const RESET_LABEL = "reset"

/*************
 * RANGE PARAMS
 *************/

const rangeParams = absVal => ({ minVal: -absVal, maxVal: absVal, stepVal: 0.01 })
const RANGES = {
    30: rangeParams(30),
    45: rangeParams(45),
    60: rangeParams(60),
    90: rangeParams(90),
    180: rangeParams(180),
}

const translateInputs = { minVal: -1, maxVal: 1, stepVal: 0.01 }

const RANGE_PARAMS = {
    dimensionInputs: { minVal: 0, maxVal: Infinity, stepVal: 1 },
    tx: translateInputs,
    ty: translateInputs,
    tz: translateInputs,
    rx: RANGES[30],
    ry: RANGES[30],
    rz: RANGES[60],
    legStance: RANGES[90],
    hipStance: RANGES[60],
    alpha: RANGES[90],
    beta: RANGES[180],
    gamma: RANGES[180],
}

/*************
 * ICONS
 *************/

const ICON_COMPONENTS = {
    mug: <GiCoffeeMug className="vertical-align" />,
    circle: <GrStatusGoodSmall className="small-icon" />,
    octocat: <FaGithubAlt className="vertical-align" />,
    check: <FaCheck className="vertical-align" />,
    times: <FaTimes className="vertical-align" />,
}

/*************
 * NAVIGATION
 *************/

const PATHS = {
    inverseKinematics: {
        path: PATH_NAMES.inverseKinematics,
        description: SECTION_NAMES.inverseKinematics,
    },
    forwardKinematics: {
        path: PATH_NAMES.forwardKinematics,
        description: SECTION_NAMES.forwardKinematics,
    },
    legPatterns: {
        path: PATH_NAMES.legPatterns,
        description: SECTION_NAMES.legPatterns,
    },
    landingPage: {
        path: PATH_NAMES.landingPage,
        description: SECTION_NAMES.landingPage,
    },
}

const KOFI_LINK_PROPERTIES = {
    name: "KOFI",
    icon: ICON_COMPONENTS.mug,
    description: "Buy Mithi Ko-Fi üçµ",
    url: "https://ko-fi.com/minimithi",
}

const REPO_LINK_PROPERTIES = {
    name: "REPO",
    icon: ICON_COMPONENTS.octocat,
    description: "Source Code",
    url: "https://github.com/mithi/hexapod",
}

const PATH_LINKS = [
    PATHS.inverseKinematics,
    PATHS.forwardKinematics,
    PATHS.legPatterns,
    PATHS.landingPage,
]

const URL_LINKS = [KOFI_LINK_PROPERTIES, REPO_LINK_PROPERTIES]

/*************
 * LANDING PAGE
 *************/

const LANDING_PAGE_MESSAGE = `

# Mithi's Bare Minimum Hexapod Robot Simulator

- Solve (and visualize) [forward][1] and [inverse][2] kinematics purely on your browser!
It's a complete rewrite of the [one][3] I wrote in Python üêç.
No more server-side computations!

- Consider buying me a [couple cups of coffee üçµ üçµ üçµ][4] to motivate me
to build other robotics related visualizers. (Quadrotors?!)

- Love coding? However big or small, any contribution to [improve the source code][5] is always appreciated. üíô

## Love, Mithi

[1]: /forward-kinematics
[2]: /inverse-kinematics
[3]: https://github.com/mithi/hexapod-robot-simulator
[4]: https://ko-fi.com/minimithi
[5]: https://github.com/mithi/hexapod/blob/master/CONTRIBUTING.md

`

export {
    ICON_COMPONENTS,
    LANDING_PAGE_MESSAGE,
    SECTION_NAMES,
    ANGLE_NAMES,
    DIMENSION_NAMES,
    LEG_NAMES,
    IK_SLIDERS_LABELS,
    RESET_LABEL,
    PATHS,
    URL_LINKS,
    PATH_LINKS,
    RANGE_PARAMS,
}
