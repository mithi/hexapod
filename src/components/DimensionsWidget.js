import React, { Component } from "react"
import NumberInputField from "./generic/NumberInputField"
import { Card, ResetButton, ToggleSwitch } from "./generic/SmallWidgets"
import { DEFAULT_DIMENSIONS } from "../templates"
import { SECTION_NAMES, DIMENSION_NAMES, RANGE_PARAMS } from "./vars"

class DimensionsWidget extends Component {
    sectionName = SECTION_NAMES.dimensions
    state = { isFine: true }

    reset = () => this.props.onUpdate("dimensions", { dimensions: DEFAULT_DIMENSIONS })

    toggleMode = () => this.setState({ isFine: !this.state.isFine })

    updateFieldState = (name, value) => this.updateDimensions(name, value)

    updateDimensions = (name, value) => {
        const dimensions = { ...this.props.params.dimensions, [name]: value }
        this.props.onUpdate("dimensions", { dimensions })
    }

    get toggleSwitch() {
        const props = {
            id: "DimensionsWidgetSwitch",
            value: this.state.isFine ? "1x" : "5x",
            handleChange: this.toggleMode,
            showValue: true,
        }

        return <ToggleSwitch {...props} />
    }

    get NumberInputFields() {
        const { minVal, maxVal } = RANGE_PARAMS.dimensionInputs
        const stepVal = this.state.isFine ? 1 : 5
        const dimensions = this.props.params.dimensions

        const numberInputFields = DIMENSION_NAMES.map(name => {
            const props = {
                name,
                value: dimensions[name],
                rangeParams: { minVal, maxVal, stepVal },
                handleChange: this.updateFieldState,
            }

            return <NumberInputField {...props} key={name} />
        })

        return <div className="grid-cols-6">{numberInputFields}</div>
    }

    render = () => (
        <Card title={<h2>{this.sectionName}</h2>} other={this.toggleSwitch}>
            {this.NumberInputFields}
            <ResetButton reset={this.reset} />
        </Card>
    )
}

export default DimensionsWidget
