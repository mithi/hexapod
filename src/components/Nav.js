import React from "react"
import { URL_LINKS, PATH_LINKS } from "./vars"
import { Link } from "react-router-dom"

const NAV_BULLETS_PREFIX = "navBullet"
const NAV_DETAILED_PREFIX = "navDetailed"

const BulletPageLink = ({ link, showDesc }) => (
    <li>
        <Link to={link.path} className="link-icon">
            <span>
                {link.icon} {showDesc ? link.description : null}
            </span>
        </Link>
    </li>
)

const BulletUrlLink = ({ path, description, icon }) => (
    <li>
        <a
            href={path}
            className="link-icon"
            target="_blank"
            rel="noopener noreferrer"
            children={
                <span>
                    {icon} {description}
                </span>
            }
        />
    </li>
)

const NavBullets = () => (
    <ul id="top-bar">
        {URL_LINKS.map(link => (
            <BulletUrlLink
                path={link.url}
                key={NAV_BULLETS_PREFIX + link.url}
                icon={link.icon}
            />
        ))}

        {PATH_LINKS.map(link => (
            <BulletPageLink key={NAV_BULLETS_PREFIX + link.path} link={link} />
        ))}
    </ul>
)

const NavDetailed = () => (
    <footer>
        <nav id="nav">
            <ul className="grid-cols-1 no-bullet">
                {URL_LINKS.map(link => (
                    <BulletUrlLink
                        path={link.url}
                        key={NAV_DETAILED_PREFIX + link.url}
                        icon={link.icon}
                        description={link.description}
                    />
                ))}

                {PATH_LINKS.map(link => (
                    <BulletPageLink
                        key={NAV_DETAILED_PREFIX + link.path}
                        link={link}
                        showDesc={true}
                    />
                ))}
            </ul>
        </nav>
    </footer>
)

const Nav = () => <NavBullets />

export { Nav, NavDetailed }
