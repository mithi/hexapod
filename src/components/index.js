import { Nav, NavDetailed } from "./Nav"
import HexapodPlot from "./HexapodPlot"
import DimensionsWidget from "./DimensionsWidget"
export { Nav, NavDetailed, HexapodPlot, DimensionsWidget }
