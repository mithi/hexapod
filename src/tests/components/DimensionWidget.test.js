import React from "react"
import { render, screen } from "@testing-library/react"
import { DimensionsWidget } from "../../components"

const DIMENSIONS = {
    front: 71,
    side: 102,
    middle: 133,
    coxia: 94,
    femur: 165,
    tibia: 154,
}

describe("Dimension Widget", () => {
    test("renders component correctly", () => {
        const { asFragment } = render(
            <DimensionsWidget params={{ dimensions: DIMENSIONS }} />
        )

        expect(asFragment()).toMatchSnapshot()

        expect(screen.getByRole("heading")).toHaveTextContent("Dimensions")
        expect(screen.getByRole("checkbox")).toBeInTheDocument()

        const button = screen.getByRole("button", { name: "reset" })
        expect(button).toBeInTheDocument()

        Object.entries(DIMENSIONS).forEach(([name, value]) => {
            const node = screen.getByRole("spinbutton", { name })
            expect(node).toHaveAttribute("value", `${value}`)
            expect(screen.getByLabelText(name)).toBeInTheDocument()
        })
    })
})
