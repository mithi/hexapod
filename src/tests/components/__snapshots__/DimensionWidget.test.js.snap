// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dimension Widget renders component correctly 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        class="row-container flex-wrap"
      >
        <h2>
          Dimensions
        </h2>
        <div
          class="switch-container"
        >
          <label
            class="switch"
          >
            <input
              type="checkbox"
              value="1x"
            />
            <span
              class="toggle-switch-widget round"
            />
            <br />
            <label
              class="label"
            >
              1x
            </label>
          </label>
        </div>
      </div>
    </div>
    <div
      class="row-container flex-wrap"
    >
      <div
        class="cell"
      >
        <label
          class="label"
          for="front"
        >
          front
        </label>
        <input
          id="front"
          inputmode="numeric"
          max="Infinity"
          min="0"
          step="1"
          style="margin: 0px;"
          type="number"
          value="71"
        />
        <label
          class="label red"
          style="opacity: 1;"
        >
           
        </label>
      </div>
      <div
        class="cell"
      >
        <label
          class="label"
          for="side"
        >
          side
        </label>
        <input
          id="side"
          inputmode="numeric"
          max="Infinity"
          min="0"
          step="1"
          style="margin: 0px;"
          type="number"
          value="102"
        />
        <label
          class="label red"
          style="opacity: 1;"
        >
           
        </label>
      </div>
      <div
        class="cell"
      >
        <label
          class="label"
          for="middle"
        >
          middle
        </label>
        <input
          id="middle"
          inputmode="numeric"
          max="Infinity"
          min="0"
          step="1"
          style="margin: 0px;"
          type="number"
          value="133"
        />
        <label
          class="label red"
          style="opacity: 1;"
        >
           
        </label>
      </div>
      <div
        class="cell"
      >
        <label
          class="label"
          for="coxia"
        >
          coxia
        </label>
        <input
          id="coxia"
          inputmode="numeric"
          max="Infinity"
          min="0"
          step="1"
          style="margin: 0px;"
          type="number"
          value="94"
        />
        <label
          class="label red"
          style="opacity: 1;"
        >
           
        </label>
      </div>
      <div
        class="cell"
      >
        <label
          class="label"
          for="femur"
        >
          femur
        </label>
        <input
          id="femur"
          inputmode="numeric"
          max="Infinity"
          min="0"
          step="1"
          style="margin: 0px;"
          type="number"
          value="165"
        />
        <label
          class="label red"
          style="opacity: 1;"
        >
           
        </label>
      </div>
      <div
        class="cell"
      >
        <label
          class="label"
          for="tibia"
        >
          tibia
        </label>
        <input
          id="tibia"
          inputmode="numeric"
          max="Infinity"
          min="0"
          step="1"
          style="margin: 0px;"
          type="number"
          value="154"
        />
        <label
          class="label red"
          style="opacity: 1;"
        >
           
        </label>
      </div>
    </div>
    <button
      class="button"
      type="button"
    >
      reset
    </button>
  </div>
</DocumentFragment>
`;
